### PR TITLE
`Plagiarism checks`: Allow students to respond to continuous plagiarism cases

### PR DESCRIPTION
**Pull request by jakubriegel**
#### General
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
Continuos plagiarism control (cpc) was introduced in #7302. It allows for automated runs of plagiarism checks and creation of plagiarism cases with a similarity warning. Those plagiarism cases were created and removed automatically when similarity was appeared or disappeared after particular check. After the exercise due date students were not able to respond to their plagiarism cases created by cpc.

### Description
This PR introduces a change which allows student  to respond to cpc plagiarism case after exercise due date (if this case was not fixed by the student before).

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 exercise with cpc enabled
- ⚠️ a version of this PR which triggers the CPC every 1 minute; you can deploy this PR for that #7443

1. Create an exercise (text, modeling or programming) with cpc enabled
    - set a future due date for the exercise
    - check the _Enable Continuos plagiarism control_ checkbox
3. Submit two identical submissions from two students.
4. Wait 1min until the plagiarism case is created.
5. Navigate to the plagiarism case from one of student accounts.
6. Verify that you cannot respond to the case.
7. Change exercise's due date to a past date.
8. Navigate to the plagiarism case from one of student accounts.
9. Verify that you can respond to the case.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| plagiarism-case-student-detail-view.component.ts | 95% | ✅                           |

### Screenshots
Plagiarism case created by cpc, before the exercise due date:
image

Plagiarism case created by cpc, after the exercise due date:
image

New option for cpc config:
image

## Summary by CodeRabbit

- **New Features**
  - Implemented a response period for students to reply to plagiarism cases within a configurable number of days.

- **Enhancements**
  - Updated plagiarism case details to include configuration settings and relevant dates.
  - Improved the student view for plagiarism cases with additional information and warning messages based on due dates.

- **Documentation**
  - Added tooltips and labels for new configuration settings in the plagiarism detection interface.

- **Bug Fixes**
  - Corrected the data type used in the plagiarism configuration test to ensure accuracy.

- **Tests**
  - Added tests to validate the new due date comparison logic in the student plagiarism case view.

@coderabbitai ignore

### CODE REVIEW

**Review by vinceclifford — APPROVED** [review]
Looks good after fixing the bug during a test session on October 25th

**Review by florian-glombik — COMMENTED** [review]
**I am unsure about the scope of this PR**
The banner was displayed as expected for a detected plagiarism case.
I am not quite sure why the banner is not displayed anymore once a student submitted a post, as multiple posts can be submitted? (I was not able to verify that with a student account as the plagiarism was not detected for some reason, but only for an admin during the testing session, so correct me if students can only submit one response)

### Change suggestions
My opinion is not too strong, but there are some things that I would change:

* I think that it is unintuitive that the textbox appears before the banner when clicking the button `Reply to Accusation`
image
* Is it intended that you can submit multiple messages?
* Is it intended that students can delete and change their messages? 
I do not think that after a certain grace period (correcting typos, ...), it is necessary to make further changes and might lead to inconsistent behaviour.
A student could change the Answer after receiving a reply, which means extra work for instructors to double check that this does not happen. 

Probably out of scope, but when testing on ts1 with these settings 
image
the plagiarism was not detected after the due date.

See e.g. the submissions of test_user_1 and test_user_2 for this exercise https://artemis-test1.artemis.cit.tum.de/courses/4/exercises/6

**Comment by jakubriegel** [comment]
> **I am unsure about the scope of this PR** The banner was displayed as expected for a detected plagiarism case. I am not quite sure why the banner is not displayed anymore once a student submitted a post, as multiple posts can be submitted? (I was not able to verify that with a student account as the plagiarism was not detected for some reason, but only for an admin during the testing session, so correct me if students can only submit one response)
> 
> ### Questions / Change suggestions
> My opinion is not too strong, but there are some things that I would change:
> 
> * I think that it is unintuitive that the textbox appears before the banner when clicking the button `Reply to Accusation`
>   image
> * Is it intended that you can submit multiple messages?
> * Is it intended that students can delete and change their messages?
>   I do not think that after a certain grace period (correcting typos, ...), it is necessary to make further changes and might lead to inconsistent behaviour.
>   A student could change the Answer after receiving a reply, which means extra work for instructors to double check that this does not happen.
> 
> Probably out of scope, but when testing on ts1 with these settings image the plagiarism was not detected after the due date.
> 
> See e.g. the submissions of test_user_1 and test_user_2 for this exercise https://artemis-test1.artemis.cit.tum.de/courses/4/exercises/6

Thanks for an in-depth review!

This PR does not change the process of students responding to plagiarism cases. It just enables the existing process for plagiarism cases created by cpc.

To answer your questions:
- _I think that it is unintuitive that the textbox appears before the banner when clicking the button `Reply to Accusation`_ **This view uses the shared component for responding to a post, which cannot be divided. That's why the banner is displayed below. Nothing was changed there in the scope of this PR.**
- _Is it intended that you can submit multiple messages?_ **That's how it is currently implemented and used for classical plagiarism cases. Nothing was changed there in the scope of this PR.**
- _Is it intended that students can delete and change their messages?_ **That's how it is currently implemented and used for classical plagiarism cases. Nothing was changed there in the scope of this PR.**

If you think that the process of responding plagiarism cases should be improved, feel free to create an issue for that. You can also talk to @MarkusPaulsen - maintainer of this domain.

About the issue with post due date checks: I've checked the link you pasted and the plagiarism case is created for this submission, so everything looks fine.

**Review by florian-glombik — DISMISSED** [review]
Thanks for the quick response and sorry for the extra effort I caused due to my misunderstanding of the scope of the PR!

As the option to respond to plagiarism accusations was displayed in the testing session it works fine!

File: src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html:52]:
```suggestion
                                !this.posts[0].answers?.length &&
```

File: src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html:64]: This timeframe needs to be adjustable by an instructor. If the due date is right before Christmas, you cannot expect  students to respond within 7 days.
With the posts it was not such a big problem, since the posts could be created on dates that do not leave to such a problem.

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html
Comment (jakubriegel) [reply src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html:67]: Thanks for your suggestion. I've added as an option to configure per exercise (screenshot attached to the description)

Comment (JohannesStoehr) [reply src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html:67]: Shouldn't you also use the newly added variable in the server to check if a student submitted the response within the correct timeframe?
And the UI also does not change after the time for a response is over, does it?

File: src/main/webapp/i18n/de/plagiarism.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/plagiarism.json:40]:
```suggestion
            "plagiarismCaseResponsePeriod": "Antwortfrist auf Plagiatsfälle in Tagen",
```

File: src/main/webapp/i18n/de/plagiarism.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/plagiarism.json:41]:
```suggestion
            "plagiarismCaseResponsePeriodTooltip": "Die Anzahl der Tage, die ein Studierender Zeit hat auf einen durch cpc generierten Plagiatsfall zu antworten, falls dieser nach der Einreichungsfrist der Aufgabe weiterhin besteht",
```

Comment (jakubriegel) [reply src/main/webapp/i18n/de/plagiarism.json:41]: Update DE copy

File: src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html
Comment (jakubriegel) [reply src/main/webapp/app/course/plagiarism-cases/student-view/detail-view/plagiarism-case-student-detail-view.component.html:67]: Response submission is not block and  UI does not change after the deadline. This is the behaviour of the original feature currently deployed on production.

I suggest we make an issue about that and implement the improvement in separate PR, so that the cpc project keeps going. What do you think?

**Review by JanaNF — COMMENTED** [review]
Old comment of mine when I ran in some issues testing it (reposting it because I wrote with Jakub directly on Slack). Maybe in the future some tester finds it helpful.

_I followed your instructions till the bullet point `3. Wait 1 min until the plagiarism case is created`. I did that and the case was never created. Or at least I never saw one as one of the students (on both server) and neither as an instructor. As an instructor I had to got to `Plagiarism Detection` and run manually the detection. Then Artemis found the plagiarism. I thought after that I could reload the page as a student and see the plagiarism case - which I did not. I also tried to confirm the plagiarism as an instructor - didn't work either._

**Comment by jakubriegel** [comment]
> Old comment of mine when I ran in some issues testing it (reposting it because I wrote with Jakub directly on Slack). Maybe in the future some tester finds it helpful.
> 
> _I followed your instructions till the bullet point `3. Wait 1 min until the plagiarism case is created`. I did that and the case was never created. Or at least I never saw one as one of the students (on both server) and neither as an instructor. As an instructor I had to got to `Plagiarism Detection` and run manually the detection. Then Artemis found the plagiarism. I thought after that I could reload the page as a student and see the plagiarism case - which I did not. I also tried to confirm the plagiarism as an instructor - didn't work either._

Thanks for testing this PR. In order to test this feature you need to configure the CPC so that it runs every 1min. You can deploy this PR to achieve it: #7443.

I've updated the description accordingly.

**Review by JanaNF — CHANGES_REQUESTED** [review]
Manual tested on legacy ts2. Worked mostly as expected.

1) Saw as instructor the plagiarism case under "Plagiarism Detection"
2) Saw as a student the "Significant Similarity" button after reloading the page of the used exercise. Could click on the button and saw the "Plagiarism Case" page, where I could not respond to the case (as expected).
3) Changed due date. Reloaded page of student and could replay (and edit the message later) to the case.
4) As instructor I saw in the "Plagiarism Detection" tab that the plagiarism case was already confirmed without me doing anything (green "Submit" button was selected). I didn't see the defend/response-message of the student. I checked the overview page of the used exercises and "Plagiarism Detection" page. When I tried to assess the submission of the student I didn't see the defend-message of the student either. 

Please let me know where I should see the message and/or if this even relevant for this PR.

**Comment by jakubriegel** [comment]
> Manual tested on legacy ts2. Worked mostly as expected.
> 
> 1. Saw as instructor the plagiarism case under "Plagiarism Detection"
> 2. Saw as a student the "Significant Similarity" button after reloading the page of the used exercise. Could click on the button and saw the "Plagiarism Case" page, where I could not respond to the case (as expected).
> 3. Changed due date. Reloaded page of student and could replay (and edit the message later) to the case.
> 4. As instructor I saw in the "Plagiarism Detection" tab that the plagiarism case was already confirmed without me doing anything (green "Submit" button was selected). I didn't see the defend/response-message of the student. I checked the overview page of the used exercises and "Plagiarism Detection" page. When I tried to assess the submission of the student I didn't see the defend-message of the student either.
> 
> Please let me know where I should see the message and/or if this even relevant for this PR.

Points 1-3 from your report look fine. Regarding point 4, I tested it again on TS4 and everything also looks fine:

Student view after responding to a case:
image

Instructor view after responding to a case:
image

**Review by JanaNF — APPROVED** [review]

**Review by dearjasmina — APPROVED** [review]
tested in a testing session, works as expected [apart from the duplicated text but we commented that on the already created issue for it]

**Review by egekurt123 — DISMISSED** [review]
Tested in testing session, works as expected. Only issue is text comparison is displayed 2000 times: issue already exists: https://github.com/ls1intum/Artemis/issues/6579

PR 7442

**Review by egekurt123 — DISMISSED** [review]

**Review by JanaNF — COMMENTED** [review]
Tried to test on legacy_ts11 and legacy_ts2 but I couldn’t connect to the server.

I wrote with Laurenz and he said:
"A new config variable was introduced by a PR, but this is not included in the config for TS11. This is why it always fails at startup. I think the PR was merged without regard to the config change and is now causing problems on all test servers where the config has not yet been adjusted. The PR with the config change is already in Develop and I suspect that if your PRs are running with the latest changes from Develop, then there will be problems"

I think your PR is affected.

**Review by JanaNF — APPROVED** [review]
Manual re-tested on legacy_ts2. Worked as expected.

**Review by laurenzfb — DISMISSED** [review]
Tested in testing session. Worked as expected

**Review by basak-akan — APPROVED** [review]
Demonstrated during the testing session, worked as expected

**Review by sarpsahinalp — DISMISSED** [review]
Demonstrated during the testing session, worked as expected

**Review by vinceclifford — DISMISSED** [review]
Works as expected. Tested during testing session

**Review by basak-akan — APPROVED** [review]
Reapprove after merge conflict resolved

**Review by laurenzfb — APPROVED** [review]
Reapprove after changes

**Review by JanaNF — APPROVED** [review]

**Review by JanaNF — APPROVED** [review]

**Review by basak-akan — APPROVED** [review]
reapprove after merge conflict resolved

**Review by laurenzfb — APPROVED** [review]
reapprove after resolving conflicts

**Review by egekurt123 — APPROVED** [review]
reapprove after resolving merge conflicts

**Comment by jakubriegel** [comment]
Conflict resolved

**Review by basak-akan — DISMISSED** [review]
Reapprove after recent code changes

**Review by egekurt123 — DISMISSED** [review]
Reapprove

**Review by laurenzfb — DISMISSED** [review]
reapprove after changes

**Review by MarkusPaulsen — DISMISSED** [review]
Code looks good and tested on TS1.

**Pull request merged**
